internal/provider/resources/resource_api_token.go

```diff
@@ -0,0 +1,566 @@
+}
+
+func (r *ApiTokenResource) ImportState(
+	ctx context.Context,
+	req resource.ImportStateRequest,
+	resp *resource.ImportStateResponse,
+) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+func (r *ApiTokenResource) ValidateConfig(
+	ctx context.Context,
+	req resource.ValidateConfigRequest,
+	resp *resource.ValidateConfigResponse,
+) {
+	var data models.ApiTokenResource
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Convert Terraform set of roles to API token roles
+	roles, diags := RequestApiTokenRoles(ctx, data.Roles)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	tokenRole, _, diags := RequestApiTokenRole(roles, data.Type.ValueString())
+	if diags != nil {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	entityType := data.Type.ValueString()
+
+	// Check if the role is valid for the entity type
+	if !utils.CheckRole(tokenRole, entityType) {
+		resp.Diagnostics.AddError(
+			"Client Error",
+			fmt.Sprintf("Role %s is not valid for entity type %s", tokenRole, entityType),
+		)
+		return
+	}
+
+	// Validate the roles based on the entity type
+	switch entityType {
+	case string(iam.ApiTokenRoleEntityTypeORGANIZATION):
+		resp.Diagnostics.Append(validateOrganizationApiToken(roles)...)
+	case string(iam.ApiTokenRoleEntityTypeWORKSPACE):
+		resp.Diagnostics.Append(validateWorkspaceApiToken(roles)...)
+	case string(iam.ApiTokenRoleEntityTypeDEPLOYMENT):
+		resp.Diagnostics.Append(validateDeploymentApiToken(roles)...)
+	}
+}
+
+func validateOrganizationApiToken(roles []iam.ApiTokenRole) diag.Diagnostics {
+	for _, role := range roles {
+		if utils.CheckRole(role.Role, string(role.EntityType)) {
+			return nil
+		}
+	}
+	return diag.Diagnostics{
+		diag.NewErrorDiagnostic(
+			"Client Error",
+			"Unable to find the role for the entity type",
+		),
+	}
+}
+
+func validateWorkspaceApiToken(roles []iam.ApiTokenRole) diag.Diagnostics {
+	for _, role := range roles {
+		if role.EntityType == iam.ApiTokenRoleEntityTypeORGANIZATION {
+			return diag.Diagnostics{
+				diag.NewErrorDiagnostic(
+					"Client Error",
+					"API Token of type WORKSPACE cannot have an ORGANIZATION role",
+				),
+			}
+		}
+
+		if utils.CheckRole(role.Role, string(role.EntityType)) {
+			return nil
+		}
+	}
+	return diag.Diagnostics{
+		diag.NewErrorDiagnostic(
+			"Client Error",
+			"Unable to find the role for the entity type",
+		),
+	}
+}
+
+func validateDeploymentApiToken(roles []iam.ApiTokenRole) diag.Diagnostics {
+	for _, role := range roles {
+		if role.EntityType != iam.ApiTokenRoleEntityTypeDEPLOYMENT {
+			return diag.Diagnostics{
+				diag.NewErrorDiagnostic(
+					"Client Error",
+					"API Token of type DEPLOYMENT cannot have an ORGANIZATION or WORKSPACE role",
+				),
+			}
+		}
+
+		if utils.CheckRole(role.Role, string(role.EntityType)) {
```

Comment:
we shouldn't be using utils/test in code here
if we want to use it, we should move it out
we should also rename this function to be more explicit, something like `ValidateRoleMatchesEntityType`

Reply:
u could prob just call this in `ValidateConfig` since all three helpers use it the same way